### PR TITLE
contracts-bedrock: ecotone gas config support

### PR DESCRIPTION
**Pull request by tynes**
**Description**

Updates the gas config on the `SystemConfig` to be ecotone first.
This resolves chain operator ux issues with setting the gas config,
rendering the additional tooling for calculating the encoded
scalar version obsolete. This version of the system config
is technically backwards compatible with pre-ecotone, but it
requires a follow up tx to call the old `setGasConfig` method.
The old gas config method was updated such that it cannot
be used to set ecotone style config to prevent footguns.
A new method `setGasConfigEcotone` should be used instead.

### CODE REVIEW

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (sebastianst) [inline packages/contracts-bedrock/src/L1/SystemConfig.sol:392]: Wondering if we explicitly also want to nullify the overhead, in case this is called on a pre-Ecotone-deployed contract. 
```suggestion
        overhead = 0;
        scalar = (uint256(0x01) << 248) | (uint256(_blobbasefeeScalar) << 32) | _basefeeScalar;
```

File: packages/contracts-bedrock/scripts/ChainAssertions.sol
Comment (sebastianst) [inline packages/contracts-bedrock/scripts/ChainAssertions.sol:74]: Do we also want to assert here that the most significant byte is `1`? We do this in the `else` branch (where the other scalar values are 0).

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:392]: I have a slight preference for not zeroing out the overhead as it gives us flexibility in passing arbitrary data into L2 if we ever need to do a "modification of interpretation of data" sort of upgrade like ecotone again, we can do this with a multicall of the legacy setter and the new setter so 2 config update events end up being emitted and we have complete control over the overhead

File: packages/contracts-bedrock/scripts/ChainAssertions.sol
Comment (tynes) [reply packages/contracts-bedrock/scripts/ChainAssertions.sol:74]: This is a good idea, thanks. I was a bit lazy here

**Comment by tynes** [comment]
@sebastianst I have addressed all open comments on this PR

File: packages/contracts-bedrock/deploy-config/devnetL1-template.json
Comment (sebastianst) [inline packages/contracts-bedrock/deploy-config/devnetL1-template.json:37]: couldn't we delete these?

File: packages/contracts-bedrock/scripts/DeployConfig.s.sol
Comment (sebastianst) [inline packages/contracts-bedrock/scripts/DeployConfig.s.sol]: same, does it even make sense to read these in still?

File: packages/contracts-bedrock/test/L1/SystemConfig.t.sol
Comment (sebastianst) [inline packages/contracts-bedrock/test/L1/SystemConfig.t.sol:329]: This works, but are weird values to use for the base fee and blob base fee scalars, as they're nominally the same values as `overhead` and `scalar` before. Maybe we use the current OP mainnet values instead?

**Review by sebastianst — APPROVED** [review]
LGTM, a few non-blocking possible improvements and cleanups

File: packages/contracts-bedrock/test/L1/SystemConfig.t.sol
Comment (tynes) [reply packages/contracts-bedrock/test/L1/SystemConfig.t.sol:329]: I am down for using whatever values here, i don't think the values themselves are super important. One consideration is that it should be a fuzz test when the values aren't super important, but I thought that would be a bit overkill in this particular situation

File: packages/contracts-bedrock/scripts/DeployConfig.s.sol
Comment (tynes) [reply packages/contracts-bedrock/scripts/DeployConfig.s.sol]: I was concerned with backwards compatibility but we could just say that the software no longer supports the older values per this commit which i think is actually how we want to do it to reduce tech debt

**Comment by maurelian** [comment]
Seeing failures in the `SystemConfig_Initialize_TestFail` and `SystemConfig_Initialize_Test` suites, possible related to the conflict in differential-testing.go?

**Comment by tynes** [comment]
> Seeing failures in the `SystemConfig_Initialize_TestFail` and `SystemConfig_Initialize_Test` suites, possible related to the conflict in differential-testing.go?

I have rebased

File: op-chain-ops/genesis/config.go
Comment (tynes) [inline op-chain-ops/genesis/config.go]: This is worth getting the eyes of the superchain registry maintainers on, cc @geoknee. The idea is that now there is no need to track the overhead and its always 0 for chains that spin up with ecotone at genesis

**Review by sebastianst — APPROVED** [review]
`scalar` and `overhead` removal lgtm! we can always add back if necessary.

**Comment by tynes** [comment]
> `scalar` and `overhead` removal lgtm! we can always add back if necessary.

Looks like `op-e2e` depends on pre-ecotone execution, meaning that we cannot fully get rid of these without a large refactor

File: op-chain-ops/genesis/config.go
Comment (tynes) [reply op-chain-ops/genesis/config.go]: I have reverted this change, no need to look into now

**Comment by tynes** [comment]
The legacy setter will need to be called for tests that don't run with ecotone enabled

**Review by trianglesphere — APPROVED** [review]

**Comment by tynes** [comment]
I have addressed merge conflicts with a rebase

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [inline packages/contracts-bedrock/src/L1/SystemConfig.sol:136]: This needs a version update now, lost it after rebase

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (mds1) [inline packages/contracts-bedrock/src/L1/SystemConfig.sol:94]: Since this is deprecated, will be renamed and replaced with a spacer at some point?

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (mds1) [inline packages/contracts-bedrock/src/L1/SystemConfig.sol:99]: Why do we use the most significant byte here instead of adding a new variable?

Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:99]: This was done as part of ecotone to enable the upgrade with no diff to the L1 contracts. The semantics of the 32 byte scalar are interpreted differently

Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:99]: See https://github.com/ethereum-optimism/specs/blob/362db2981471d3d96871a44985789e838653ee22/specs/protocol/system_config.md#ecotone-scalar-overhead-uint256uint256-change

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (mds1) [inline packages/contracts-bedrock/src/L1/SystemConfig.sol]: Since this is the first contract change towards the future `op-contracts/v1.6.0` release, what do you think about versioning this as `2.3.0-beta.1`? I plan to write up a design doc for a new way to do contract semvers, which basically will be:
- `X.Y.Z-beta.n` during development
- `X.Y.Z-rc.n` once it becomes a release candidate
- `X.Y.Z` once it becomes an official release

The proposal would also support appending `+feature` for compatibility with https://github.com/ethereum-optimism/design-docs/pull/8

So in this case specifically, assuming the SystemConfig will have no breaking changes before the next release, this PR becomes 2.3.0–beta.1,the next PR becomes 2.3.0-beta.2, and we continue incrementing in that fashion until we get to the release candidate version

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:94]: I didn't want to delete it and turn it into a spacer in case we need to use it as part of a weird upgrade mechanism as the value is still passed into an event and picked up by `op-node` but its a noop

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol]: I am in favor of this. I would like to update the table in https://github.com/ethereum-optimism/optimism/issues/10592 to reflect this scheme. Can update the version to be this

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (mds1) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:94]: Could you expand on this a bit more? I see that it's passed into the event but don't follow the potential upgrade use case. (Also not a blocker for this PR, and we can track this decision separately if needed)

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol]: I updated the table to be broken out to list of the versions of specific contracts we want to release and then a changelog sort of section for each contract

File: packages/contracts-bedrock/test/L1/SystemConfig.t.sol
Comment (mds1) [inline packages/contracts-bedrock/test/L1/SystemConfig.t.sol]:
```suggestion
    /// @dev Tests that initialization sets the correct values.
```

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (mds1) [inline packages/contracts-bedrock/src/L1/SystemConfig.sol:369]: How do you feel about pulling out the `_scalar` decoding logic into internal helper methods? Arguably overkill since it's one line, but it lets us cleanly fuzz test the decoding (parsing this way vs. with foundry cheatcodes) to ensure it conforms to the spec

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (mds1) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol]: https://github.com/ethereum-optimism/optimism/issues/10592 looks really great, thanks a lot for setting that up

File: packages/contracts-bedrock/scripts/ChainAssertions.sol
Comment (mds1) [inline packages/contracts-bedrock/scripts/ChainAssertions.sol:79]: Best practice with `require` statements IMO is to always use the string argument since it makes it a lot easier to identify where exactly an assertion failed. This is similar to what we did in the postCheck of superchain ops. I don't want to block this PR on modifying old code, but would at least like to start migrating new code to that. 

I like incrementing numbers of 100 so you can add new ones in between without adjusting all numbers, for example:
```solidity
require(config.unsafeBlockSigner() == _cfg.p2pSequencerAddress(), "600");
require(config.scalar() >> 248 == 1, "700");
```

then in the future, minimal diff to add a new check
```solidity
require(config.unsafeBlockSigner() == _cfg.p2pSequencerAddress(), "600");
require(myNewCheck, "650");
require(config.scalar() >> 248 == 1, "700");
```

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:369]: This function still exists to support pre ecotone chains. There is no decoding here - this enforcement that the most significant byte is not set is just UX sugar. Its basically saying "use the `setGasConfig` interface if you are pre ecotone and use `setGasConfigEcotone` if you are post ecotone

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:94]: See discussion here: https://github.com/ethereum-optimism/optimism/pull/10226/files#r1575444717

Basically if we ever upgrade L2 without upgrading L1, we now have extra data that we can pass into L2. This is helpful if for some reason there is a constraint that we cannot upgrade the L1 contracts in a timely manner

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:136]: Version has been updated

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:369]: Ah perhaps you meant to leave this comment in the other function where it is encoded?

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (mds1) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:94]: Got it, thanks

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:369]: For what its worth, there is e2e differential fuzzing of the solidity code that generates the scalar value, it doesn't directly operate over a library which you are right that is a thinner abstraction but it does directly fuzz over the abstraction that the `op-node` consumes

Comment (mds1) [reply packages/contracts-bedrock/src/L1/SystemConfig.sol:369]: Got it, that sounds sufficient to me

**Review by mds1 — APPROVED** [review]

File: packages/contracts-bedrock/scripts/ChainAssertions.sol
Comment (tynes) [reply packages/contracts-bedrock/scripts/ChainAssertions.sol:79]: definitely agree with this, at some point we should do this

**Pull request merged**
